File: tools/skill-loader-py/examples/load_skill.py
from __future__ import annotations

import argparse
import json
import sys
from pathlib import Path

# Allow running this script without installing the project.
_ROOT = Path(__file__).resolve().parents[1]
if str(_ROOT) not in sys.path:
    sys.path.insert(0, str(_ROOT))

from fetch import fetch_skill  # noqa: E402
from loader import Registry, load_and_register  # noqa: E402
from verify import compute_tree_sha256, verify_pinned_ref, verify_sha256  # noqa: E402


def main() -> int:
    ap = argparse.ArgumentParser()
    ap.add_argument("--local", help="Load from a local directory (no fetch)")
    ap.add_argument("--owner")
    ap.add_argument("--repo")
    ap.add_argument("--ref", default=None)
    ap.add_argument("--method", default="git", choices=["git", "zip"])
    ap.add_argument("--skills-dir", default="./external_skills")
    ap.add_argument("--expected-commit", default=None)  # set your pinned commit here
    ap.add_argument("--expected-tag", default=None)  # or set pinned tag here
    ap.add_argument("--expected-sha256", default=None)
    ap.add_argument("--dry-run", action="store_true")
    args = ap.parse_args()

    registry = Registry.create()

    if args.local:
        skill_dir = str(Path(args.local).resolve())
    else:
        if not args.owner or not args.repo:
            raise SystemExit("Provide --owner and --repo, or use --local")

        ref = args.ref or args.expected_commit or args.expected_tag

        if args.dry_run:
            print(
                json.dumps(
                    {
                        "dryRun": True,
                        "wouldFetch": {
                            "owner": args.owner,
                            "repo": args.repo,
                            "ref": ref,
                            "method": args.method,
                            "skillsDir": args.skills_dir,
                        },
                    },
                    indent=2,
                )
            )
            return 0

        fetched = fetch_skill(
            owner=args.owner,
            repo=args.repo,
            dest_dir=args.skills_dir,
            ref=ref,
            method=args.method,
            clean=True,
        )
        skill_dir = fetched.skill_dir

    if args.expected_commit or args.expected_tag:
        pinned = verify_pinned_ref(skill_dir, expected_commit=args.expected_commit, expected_tag=args.expected_tag)
        if not pinned.ok:
            raise SystemExit(f"Pinned ref verification failed: {pinned.reason}")

    actual_sha = compute_tree_sha256(skill_dir)
    if args.expected_sha256:
        r = verify_sha256(skill_dir, args.expected_sha256)
        if not r["ok"]:
            raise SystemExit(f"sha256 mismatch: expected {r['expected']} got {r['actual']}")

    reg = load_and_register(skill_dir=skill_dir, registry=registry, runtime="python")

    print(
        json.dumps(
            {
                "registered": reg,
                "computed": {"sha256": actual_sha},
            },
            indent=2,
        )
    )
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: tools/skill-loader-node/examples/load-skill.js
import path from 'node:path';
import process from 'node:process';
import { fetchSkill } from '../src/fetch.js';
import { createRegistry, loadAndRegister } from '../src/loader.js';
import { computeTreeSha256, verifyPinnedRef, verifySha256 } from '../src/verify.js';

function arg(name) {
  const i = process.argv.indexOf(name);
  return i >= 0 ? process.argv[i + 1] : null;
}

function flag(name) {
  return process.argv.includes(name);
}

async function main() {
  const registry = createRegistry();

  const localDir = arg('--local');
  const skillsDir = arg('--skills-dir') ?? './external_skills';

  const expectedCommit = arg('--expected-commit') ?? null; // set your pinned commit here
  const expectedTag = arg('--expected-tag') ?? null; // or set pinned tag here
  const expectedSha256 = arg('--expected-sha256') ?? null; // optional

  let skillDir;

  if (localDir) {
    skillDir = path.resolve(localDir);
  } else {
    const owner = arg('--owner');
    const repo = arg('--repo');
    const ref = arg('--ref') ?? expectedCommit ?? expectedTag ?? null;
    const method = arg('--method') ?? 'git';

    if (!owner || !repo) {
      throw new Error('Provide --owner and --repo, or use --local');
    }

    if (flag('--dry-run')) {
      console.log('[dry-run] would fetch skill', { owner, repo, ref, method, skillsDir });
      process.exit(0);
    }

    const fetched = await fetchSkill({ owner, repo, destDir: skillsDir, ref, method, clean: true });
    skillDir = fetched.skillDir;
  }

  // Integrity checks (require local git checkout for pinned ref checks).
  if (expectedCommit || expectedTag) {
    const pinned = await verifyPinnedRef(skillDir, { expectedCommit, expectedTag });
    if (!pinned.ok) throw new Error(`Pinned ref verification failed: ${pinned.reason}`);
  }

  const actualSha256 = await computeTreeSha256(skillDir);
  if (expectedSha256) {
    const r = await verifySha256(skillDir, expectedSha256);
    if (!r.ok) throw new Error(`sha256 mismatch: expected ${r.expected} got ${r.actual}`);
  }

  const registration = await loadAndRegister({ skillDir, registry, runtime: 'node' });

  console.log(
    JSON.stringify(
      {
        registered: registration,
        computed: { sha256: actualSha256 }
      },
      null,
      2
    )
  );
}

main().catch((err) => {
  console.error(err?.stack || String(err));
  process.exit(1);
});
